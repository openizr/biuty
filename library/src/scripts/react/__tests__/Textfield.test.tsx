/**
 * Copyright (c) Matthieu Jabbour. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import UITextfield from 'scripts/react/Textfield';
import { act, Simulate } from 'react-dom/test-utils';
import { render, unmountComponentAtNode } from 'react-dom';

type Target = any; // eslint-disable-line @typescript-eslint/no-explicit-any

let container = document.createElement('div');
jest.mock('scripts/helpers/generateRandomId');

describe('react/UITextfield', () => {
  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
    jest.clearAllMocks();
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    ((container as unknown) as null) = null;
  });

  test('renders correctly - basic', () => {
    act(() => {
      render(<UITextfield name="test" modifiers="large" />, container);
    });
    const input = document.querySelector('input') as HTMLInputElement;
    act(() => {
      Simulate.focus(input);
    });
    act(() => {
      Simulate.change(input, { target: { value: 'new test' } as Target });
    });
    act(() => {
      Simulate.blur(input);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with id', () => {
    act(() => {
      render(<UITextfield name="test" id="test" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with type number', () => {
    act(() => {
      render(<UITextfield name="test" type="number" min={0} max={30} step={10} />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with type text', () => {
    act(() => {
      render(<UITextfield name="test" type="text" maxlength={10} size={10} />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with placeholder', () => {
    act(() => {
      render(<UITextfield name="test" placeholder="test" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with helper', () => {
    act(() => {
      render(<UITextfield name="test" helper="test" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with label', () => {
    act(() => {
      render(<UITextfield name="test" label="test" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with value', () => {
    act(() => {
      render(<UITextfield name="test" value="test" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with left icon', () => {
    act(() => {
      render(<UITextfield name="test" icon="star" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with right icon', () => {
    act(() => {
      render(<UITextfield name="test" icon="star" iconPosition="right" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - disabled', () => {
    act(() => {
      render(<UITextfield name="test" modifiers="disabled" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - autocomplete off', () => {
    act(() => {
      render(<UITextfield name="test" autocomplete="off" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - readonly', () => {
    const onChange = jest.fn();
    act(() => {
      render(<UITextfield name="test" readonly onChange={onChange} />, container);
    });
    const input = document.querySelector('input') as HTMLInputElement;
    act(() => {
      Simulate.change(input, { target: { value: 'new test' } as Target });
    });
    expect(container).toMatchSnapshot();
    expect(onChange).toHaveBeenCalledTimes(0);
  });

  test('renders correctly - with listeners', () => {
    const onFocus = jest.fn();
    const onBlur = jest.fn();
    const onChange = jest.fn();
    const onIconClick = jest.fn();
    act(() => {
      render(
        <UITextfield
          name="test"
          icon="star"
          value="test"
          onChange={onChange}
          onFocus={onFocus}
          onBlur={onBlur}
          onIconClick={onIconClick}
        />,
        container,
      );
    });
    const icon = document.querySelector('i') as HTMLElement;
    const input = document.querySelector('input') as HTMLInputElement;
    act(() => {
      Simulate.focus(input);
    });
    act(() => {
      Simulate.change(input, { target: { value: 'new test' } as Target });
    });
    act(() => {
      Simulate.blur(input);
    });
    act(() => {
      Simulate.click(icon);
    });
    expect(container).toMatchSnapshot();
    expect(onFocus).toHaveBeenCalledTimes(1);
    expect(onFocus).toHaveBeenCalledWith('test');
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith('new test');
    expect(onBlur).toHaveBeenCalledTimes(1);
    expect(onBlur).toHaveBeenCalledWith('new test');
    expect(onIconClick).toHaveBeenCalledTimes(1);
  });
});
