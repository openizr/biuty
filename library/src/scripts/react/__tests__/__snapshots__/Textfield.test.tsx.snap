// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react/UITextfield renders correctly - autocomplete off 1`] = `
<div>
  <div
    class="ui-textfield"
  >
    <div
      class="ui-textfield__wrapper"
    >
      <input
        autocomplete="off"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextfield renders correctly - basic 1`] = `
<div>
  <div
    class="ui-textfield ui-textfield--large"
  >
    <div
      class="ui-textfield__wrapper"
    >
      <input
        autocomplete="on"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        type="text"
        value="new test"
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextfield renders correctly - disabled 1`] = `
<div>
  <div
    class="ui-textfield ui-textfield--disabled"
  >
    <div
      class="ui-textfield__wrapper"
    >
      <input
        autocomplete="on"
        class="ui-textfield__wrapper__field"
        disabled=""
        id="z8a8d7a5zad4"
        name="test"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextfield renders correctly - readonly 1`] = `
<div>
  <div
    class="ui-textfield"
  >
    <div
      class="ui-textfield__wrapper"
    >
      <input
        autocomplete="on"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        readonly=""
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextfield renders correctly - with helper 1`] = `
<div>
  <div
    class="ui-textfield"
  >
    <div
      class="ui-textfield__wrapper"
    >
      <input
        autocomplete="on"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        type="text"
        value=""
      />
    </div>
    <span
      class="ui-textfield__helper"
    >
      test
    </span>
  </div>
</div>
`;

exports[`react/UITextfield renders correctly - with id 1`] = `
<div>
  <div
    class="ui-textfield"
    id="test"
  >
    <div
      class="ui-textfield__wrapper"
    >
      <input
        autocomplete="on"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextfield renders correctly - with label 1`] = `
<div>
  <div
    class="ui-textfield"
  >
    <label
      class="ui-textfield__label"
      for="z8a8d7a5zad4"
    >
      <strong>
        test
      </strong>
    </label>
    <div
      class="ui-textfield__wrapper"
    >
      <input
        autocomplete="on"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextfield renders correctly - with left icon 1`] = `
<div>
  <div
    class="ui-textfield"
  >
    <div
      class="ui-textfield__wrapper"
    >
      <i
        class="ui-textfield__wrapper__icon"
        role="button"
        tabindex="0"
      >
        star
      </i>
      <input
        autocomplete="on"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextfield renders correctly - with listener and debounce 1`] = `
<div>
  <div
    class="ui-textfield"
  >
    <div
      class="ui-textfield__wrapper"
    >
      <i
        class="ui-textfield__wrapper__icon"
        role="button"
        tabindex="0"
      >
        star
      </i>
      <input
        autocomplete="on"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        type="text"
        value="new test"
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextfield renders correctly - with listeners 1`] = `
<div>
  <div
    class="ui-textfield"
  >
    <div
      class="ui-textfield__wrapper"
    >
      <i
        class="ui-textfield__wrapper__icon"
        role="button"
        tabindex="0"
      >
        star
      </i>
      <input
        autocomplete="on"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        type="text"
        value="new test"
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextfield renders correctly - with placeholder 1`] = `
<div>
  <div
    class="ui-textfield"
  >
    <div
      class="ui-textfield__wrapper"
    >
      <input
        autocomplete="on"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        placeholder="test"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextfield renders correctly - with right icon 1`] = `
<div>
  <div
    class="ui-textfield"
  >
    <div
      class="ui-textfield__wrapper"
    >
      <input
        autocomplete="on"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        type="text"
        value=""
      />
      <i
        class="ui-textfield__wrapper__icon"
        role="button"
        tabindex="0"
      >
        star
      </i>
    </div>
  </div>
</div>
`;

exports[`react/UITextfield renders correctly - with transform 1`] = `
<div>
  <div
    class="ui-textfield"
  >
    <div
      class="ui-textfield__wrapper"
    >
      <input
        autocomplete="on"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        type="text"
        value="NEW TEST"
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextfield renders correctly - with type number 1`] = `
<div>
  <div
    class="ui-textfield"
  >
    <div
      class="ui-textfield__wrapper"
    >
      <input
        autocomplete="on"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        max="30"
        min="0"
        name="test"
        step="10"
        type="number"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextfield renders correctly - with type text 1`] = `
<div>
  <div
    class="ui-textfield"
  >
    <div
      class="ui-textfield__wrapper"
    >
      <input
        autocomplete="on"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        maxlength="10"
        name="test"
        size="10"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`react/UITextfield renders correctly - with value 1`] = `
<div>
  <div
    class="ui-textfield"
  >
    <div
      class="ui-textfield__wrapper"
    >
      <input
        autocomplete="on"
        class="ui-textfield__wrapper__field"
        id="z8a8d7a5zad4"
        name="test"
        type="text"
        value="test"
      />
    </div>
  </div>
</div>
`;
